Add support for OpenAI built-in web search tool

twang849 opened the pull request:
### What
- Added web search tool "openai_websearch.py" 
- Added tool to __init__.py so that it can be imported as "OpenAIWebSearchTool"
- Modified the OSS Deep Researcher agent to utilize this tool instead of Tavily

Closes #212

drbrady8800 reviewed (changes requested):
This looks good, I wouldn't rely on the llm to call launch model before it tries web search (it never is called in deep research so every call fails). I would just make the client every time you call perform the web search.

twang849 commented:
Thanks, makes sense, I will make the change.

drbrady8800 commented on src/agentic/tools/openai_websearch.py:
```diff
@@ -0,0 +1,135 @@
+        """Return a list of functions that will be exposed to the agent."""
+        return [
+            self.perform_web_search,
+            self.launch_model
```
This is no longer a function, can remove

drbrady8800 commented on src/agentic/tools/openai_websearch.py:
```diff
@@ -0,0 +1,135 @@
+
+from openai import OpenAI
+
+from agentic.common import Agent, AgentRunner, RunContext
```
We have switched terminology, use ThreadContext instead of RunContext

drbrady8800 commented on src/agentic/tools/openai_websearch.py:
```diff
@@ -0,0 +1,135 @@
+            self.launch_model
+        ]
+    
+    def perform_web_search(self, prompt: str,  run_context: RunContext):
```
thread_context

drbrady8800 reviewed (changes requested):
Looks good, a couple issues with naming / passing args through. Try to test deep research to make sure it works fully while you test. That is our most complex agent, so if it works, you can be reasonably sure simpler agents will too. The dedupe sources is also not working, sources are embedded in this format: [text](link) we should parse for those and use the title / url. For testing let's use `gpt-4o-mini-search-preview` it is much cheaper

twang849 commented:
@drbrady8800 Fixed a bunch of the issues that you commented in the code. For the sources, I replaced the _deduplicate_and_format_sources function with format_sources, which removes the embedded links and places them into a dictionary var called sources. However I realized that the model doesnt need the code to sort out repeated sources, so right now that function is not being activated in query_web_content. Please let me know what you think

twang849 commented:
@drbrady8800 Okay, I will take a look at it again. Also, when you say that format_sources should be a util function, do you mean prefixing it with an underscore or something else? Thanks

twang849 commented:
@drbrady8800 Okay, I moved the helper function to tools.utils.registry, let me know if that is what you meant. I also changed it so that the sources are deduplicated and listed at the end of each web search, like how it was with the Tavily tool. It seems to be working fine now, I've run it three times and it's correctly shown sources at the end. Though there is a weird bug where the Introduction is being shown after the Conclusion, I'm not sure if this has something to do with the new tool?

drbrady8800 commented on src/agentic/tools/utils/registry.py:
```diff
@@ -20,6 +21,34 @@ class ConfigRequirement:
+    sourcesString = "Sources: " + '\n'.join(sources)
+
+    result = formatted_content + '\n' + sourcesString
+    print(result)
```
Remove this print

drbrady8800 commented on src/agentic/tools/utils/registry.py:
```diff
@@ -20,6 +21,34 @@ class ConfigRequirement:
     required: bool = True
     default: Any = None
 
+def format_sources(content: str):
```
After looking more, I don't think this really belongs in this file. My mistake for directing you here. I would make a new util file in this directory just called format_sources, and just include this function in it. If you'd like you could give it a more general title like text_parsing or something like that.

drbrady8800 reviewed (approved):
Looks great! Thanks for bearing with all the changes, excited to have native openai in here. If you could just move the util and remove the print statement we should be good!
